### ISSUE TITLE
[RFC] Nomenclature of the components exported by `@mui/x-date-pickers` and `@mui/x-date-pickers-pro`

### ISSUE THREAD

**Issue by eps1lon** [mui/mui-x#4484]
TL;DR: Changes to the public API are
1. `Static*Picker` will become `*PickerView` (naming open).
1. `DayPicker` will become `CalendarPicker`
1. `DayPicker` `view` `date` will become `day`
1. Drop `allowKeyboardControl`
1. components implementing `<input type={...} />` will end in `Input` instead of `Picker` e.g. `TimePicker` will become `TimeInput`

For a short summary of the terminology used see https://gist.github.com/eps1lon/97cbf619cffd30f12deece8d72b0c941

## variants

Right now we ship every picker in 4 different variants:
1. static e.g. `StaticDatePicker`
2. mobile e.g. `MobileDatePicker`
3. desktop e.g. `DesktopDatePicker`
4. responsive e.g. `DatePicker`

Static variants are the underyling picker without the textbox(s). It isn't static at all since it takes user input, validates it etc. 
Mobile and desktop are for their respective environment and "responsive" just switches dynamically between "mobile" and "desktop" based on a media query.

None of these components share a common entry point in their implementation (anymore). However, it seems to me that all of these compose naturally into:
1. Responsive
   1. Desktop
      1. desktop textbox
      1. desktop popup
         1. popup content
   1. Mobile
      1. mobile textbox
      1. mobile popup
         1. popup content

The "popup content" is what will become the `*PickerView` component (or `*PickerGrid`, `*PickerDialogContainer`). The naming is still open since it's somewhat arbitrary what we consider the "view". It's just as arbitrary as "static" but it seems to me that "static" is even more misleading.

The "responsive" component will then just be a thin wrapper around the desktop and tooltip variants that doesn't need much customization. Instead, we should encourage custom wrappers especially for lazy loading etc.

## timing related naming

_cursive formatting_ refers to natural language which isn't clearly defined (especially considering different languages which differ in what is a _date_ or _time_).
`monospace` formatting refers to technical terms that are well defined.

- **DatePicker**: picks a `date` with textbox + calendarpicker
- **TimePicker**: picks a `time ` with textbox + clockpicker
- **DateTimePicker**: picks a `moment` with textbox + (clockpicker + calendarpicker + switch for calendar vs clock)
- **CalendarPicker**: picks `year`, `month` and `day`
- **ClockPicker**: picks `hour`, `minute` and `second`
- **view**: a part of calendar or clockpicker e.g. `year`

What we call a `date` is not the same as a JavaScript `Date`. Instead we follow the naming of HTML input types where we split a _point in time_ (a `moment`) into `date` and `time` ([ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) equally applies). `date` is further split into `year`, `month` and `day`. `time` is split into `hour`, `minute` and `second`. 

## allowKeyboardControl

Dropping `allowKeyboardControl` is more of a drive-by improvement. This is effectively a prop that makes the component strictly worse (since it excludes people relying on keyboard input). I suspect the prop was motivated by the fact that the original implementation registered all keyboard handlers globally when we can use keydown handlers on the specific components instead.
     
## `Picker` vs `Input`

This change is unlikely to land in the public interface. But it might make sense to use that terminology internally. 

Right now we use `picker` as a generic term for all components that were part of `@material-ui/pickers`. But it's unclear what this term means considering we have `ClockPicker` and `TimePicker` (which is just `ClockPicker` + textbox). 

Let's entertain the idea of renaming `DatePicker`  to `DateInput` and `CalendarPicker` to `DatePicker`. `DateInput ` composes a textbox + `DatePicker` . *Picker referring to the UI that you choose from pre-defined values (like the `listbox`  in a `combobox`). Seems like the best solution for cleaning up the terminology. But I'm afraid causing more confusion for people being used to `DatePicker`.

I would propose 
```diff
-TimePicker
+TimeInput
-DatePicker
+DateInput
-DateTimePicker
+DateTimeInput
-DateRangePicker
+DateRangeInput
```

/cc @mui-org/maintainers

**Comment by flaviendelangle** [comment]
@alexfauquette @joserodolfofreitas that one will be interesting to keep track off

**Comment by flaviendelangle** [comment]
I am restarting this discussion to be able to have something concrete in the coming week and to implement the new nomenclature before the end of the alpha.

## Current behavior

**Side note:** I am using the term *view* as all the panels that focuses on a single date section.
For most date section, 1 view = 1 panel visible on screen (we list all the months of the year in a single panel, we show all the hours of the day in a single panel etc...).
But for the `DayPicker` (to be renamed) component, we have a small inconsistency because sometime we say the `DayPicker` is a single view and sometime we say each month in the `DayPicker` is a view.

My understanding is that we have **4 families of components** 

1. The components with only an HTML input (current `DateField` and `DateRangeField`)

2. The components with an HTML input and a view displayed in a modal or a tooltip (current `DatePicker`, `MobileDatePicker`, `DesktopDatePicker`, `DateRangePicker`, ...)

3. The components with only **ONE** view rendered directly on the page (current `DayPicker`, `MonthPicker`, `YearPicker`)

4. The components with several views rendered directly on the page (current `StaticDatePicker`, `StaticTimePicker`, ...)

## Proposal

The list below covers more than just the components exposed by our package.
It also covers some slots naming because we can't decide how to name our components without deciding which names are already taken for another concept in the slots.

### Input

*Component used to render an `HTML` component*

**Examples:**

```tsx
// Set the input rendered in the `DateField` component
// Only one `OutlinedInput` will be rendered
<DateField components={{ input: OutlinedInput }} />

// Set the input rendered in the `MultiInputDateRangeField` component
// Two `FilledInput` will be rendered, one for the start date and one for the end date
<MultiInputDateRangeField components={{ input: FilledInput }} />
```

### Field

*Component used to edit a date with the keyboard*

**Examples:**
```tsx
// Standalone usage of a field component
<DateField />

// Set the field rendered in the `DateRangePicker` component
<DateRangePicker components={{ field: MultiInputDateRangeField }} />
<DateRangePicker components={{ field: SingleInputDateRangeField }} />
```

**Competitors naming for the same concept:**
- Telerik: Input
- Spectrum: Field

### View

*Component used to edit one or several date section (day, month, year, hour, minutes, seconds) with the mouse*

**Examples:**
```tsx
// Standalone usage of views
<MonthView /> // current `MonthPicker`, to edit month
<YearView /> // current `YearPicker`, to edit year
<DayView /> // current `DayPicker`, to edit day
<CalendarView /> // current `CalendarPicker`, to edit day / month / year

// Set the list of views rendered in a `DatePicker` component
<DatePicker views={['month', 'day']} />
```

**Competitors naming for the same concept:**
- Telerik: View
- Spectrum: no prefix suffix (`RangeCalendar` for instance)
- Retool: Input

### Picker

*Component used to edit a date with both the keyboard and the mouse*
*Is the combination of a field and one or several views*

**Examples:**
```tsx
// Standalone usage of picker components
<DatePicker />
<DateRangePicker />
<DateTimePicker />
<DateRangeTimePicker />
```

**Competitors naming for the same concept:**
- Telerik: Picker
- Spectrum: Picker
- Retool: No prefix / suffix (`DateTime` for instance, but I did not find an actual component name, just the doc page name)

**Comment by oliviertassinari** [comment]
@flaviendelangle sounds great. One this demo,

```jsx
<DateField components={{ input: OutlinedInput }} />
```

I think that for many teams, they will want to use:

```jsx
<DateField components={{ input: TextField }} />
```

Which I think is still coherent if we argue that a DateField is a specialized version of a TextField.

**Comment by LukasTy** [comment]
> My main goal here is to be consistent across our components (MonthPicker should clearly not have the same suffix as DatePicker for instance)

Could you elaborate more on this point? 🤔 
IMHO, I'd argue that the current naming of those components represents their intent more correctly than `View`.
`MonthView` - I might expect to see a visual representation of months in a year or days in a month given such name.
`MonthPicker` - I would expect a component with ability to pick/choose a particular month

**Comment by flaviendelangle** [comment]
>
> I think that for many teams they will want to use:
>
> ```tsx
> <DateField components={{ input: TextField }} />
> ```
>

The exact behavior of the `input` slot is something I would like to discuss, but maybe in a standalone issue.
Should it have a default value ? If so maybe not the `TextField` due to the bundle size but it should clearly be compatible with the `TextField` out of the box.
Should it remain a render prop like today ? 
Etc...

_____

> My main goal here is to be consistent across our components (MonthPicker should clearly not have the same suffix as DatePicker for instance)
>
> Could you elaborate more on this point? 

Currently, the `Picker` term refers to either:
- a UI where you have both an HTML input and a UI rendered in a modal / tooltip ( `DatePicker`, `DateRangePicker`, ...)
- a UI allowing you to select a date (or part of a date) directly on your screen (`CalendarPicker`, `MonthPicker`, `YearPicker`, or with a prefix `StaticDatePicker`, `StaticDateRangePicker`, ...) 

For me there is an inconsistency there.
And by reading the name of the component, we should know if there is an input and a modal, or if the content is directly on your screen.

We can of course discuss this naming, the most important part for me being the differentiation between the two.

For example, we could add the `Static` prefix to all of them (`StaticMonthPicker`).
But I don't have the feeling the `Static` describes the actual behavior either.

**Comment by LukasTy** [comment]
> For example, we could add the Static prefix to all of them (StaticMonthPicker).

Oh, I see your point. That is a problem we could solve by tweaking the naming and you did give a very good example of using `Static` as we already use it for other pickers without input.

> But I don't have the feeling the Static describes the actual behavior either.

Indeed, there are a few of ways we could go:
1. come up with a better substitute for `Static` as per @eps1lon suggestion (also adding this name to the components in question)
2. add `Static` to `YearPicker`, `MonthPicker`, `CalendarPicker` to align with existing `Static<Type>Picker`
3. suffix all `static` components with `View` (i.e.: `YearPicker` -> `YearPickerView`, `StaticDatePicker` -> `DatePickerView`)
4. rename the components, which work with input (`DatePicker` -> `DateInput`) as per @eps1lon suggestion

I've listed the options in my personal preference order.
I'm not a fan of the last option just because of the possible confusion caused to developers.
One possible alternatives to `Static` I could think of is `Fixed`. 🤔

**Comment by flaviendelangle** [comment]
I was thinking about the `Inline` prefix.
@alexfauquette if you have some idea.

**Comment by alexfauquette** [comment]
## About the current renaming

From What I understood from the previous discussion, the current modification proposal is the next one.

Some naming do not change. The components with an input are ending with `Field`

| v5 name | v6 name |
| --- | ---|
| `DateTimePicker` | same |
| `TimePicker` | same |
| `DateRangePicker` | same |
| `DatePicker` | same |
| |
| `DesktopDatePicker` | same |
| `DesktopDateRangePicker` | same |
| `DesktopDateTimePicker` | same |
| `DesktopTimePicker` | same |
| |
| `MobileDatePicker` | same |
| `MobileDateRangePicker` | same |
| `MobileDateTimePicker` | same |
| `MobileTimePicker` | same |

The components that are referring to a `views=['day', 'year']` end with `View` for consistency

| v5 name | v6 name |
| --- | ---|
| `MonthPicker` | `MonthView`|
| `YearPicker` | `YearView`|
| `DateRangePickerDay` | `MultiMonthDayView` (tricky one)|
| `PickersDay` | `DayView`|


For Clock and Calendar I have a doubt. Should we emphasize that it's an aggregation of different views. Interface si conflicting with the notion of interface of typescript

| v5 name | v6 name |
| --- | ---|
| `ClockPicker` | `ClockView` or `ClockInterface` |
| `CalendarPicker` | `CalendarView` or `CalendarInterface` |

Same question for statics

| v5 name | v6 name |
| --- | ---|
| `StaticTimePicker` | `TimePickerView` or `TimePickerInterface` |
| `StaticDateRangePicker` | `DateRangePickerView` or `DateRangePickerInterface` |
| `StaticDateTimePicker` | `DateTimePickerView` or `DateTimePickerInterface` |
| `StaticDatePicker` | `DatePickerView` or `DatePickerInterface` |

We also export `CalendarPickerSkeleton` and `PickerStaticWrapper` be nobody care and they are not a problem

### The new components

From a developer point of view (I'm not sure about how complex it's to implement) I would expect `DateField` to behave like `TextField`. So you no not pass `TextField` to its slots, but the subcomponents

```jsx
// Simple customization
<DateField label="..." placeholder="..." components={{ 
    Input: OutlinedInput,
	InputLabel: ..
    FormHelperText: ... 
}} />
```

----

@flaviendelangle I don't get for wichi component do you want to pass `Inline` For components without views?

**Comment by flaviendelangle** [comment]
> @flaviendelangle I don't get for wichi component do you want to pass Inline For components without views?

The debate is about those components:

> The components that are referring to a views=['day', 'year'] end with View for consistency

@LukasTy is not a fan of the `View` suffix, and I agree that it's worth taking some time to think about it.

____

> We also export CalendarPickerSkeleton and PickerStaticWrapper be nobody care and they are not a problem

Yes for those two I don't really understand the use case yet.

____

> From a developer point of view (I'm not sure about how complex it's to implement) I would expect DateField to behave like TextField.

The problem being that on a date range field, we can have 2 `TextField` rendered.
In a perfect world, I would have a suffix for `DateField` & co that is not `Field` but something never used.

**Comment by joserodolfofreitas** [comment]
> In a perfect world, I would have a suffix for DateField & co that is not Field but something never used.

I think "Field" makes sense, although, as you said is the closest but not ideal.

Personally, I see everything as an input, you can input the date with a mouse (using a picker) or with a keyboard (using a field). But I agree that input is usually related to the most basic component of the hierarchy.

The naming that would make the most sense to me would be:

- Advanced component composed of fields and pickers: No prefix.
- - Mouse/touch inputs -> Prefix: Picker
- - Keyboard inputs -> Prefix: Field
- - - HTML inputs -> Prefix: Input

Example:
`<DateRange pickers: ['month', 'year']  components: { field: MultiInputDateRangeField, monthPicker: MonthPicker } >`

**Comment by flaviendelangle** [comment]
So in your nomenclature, if I understand correctly we would have:

- `DatePicker` => `Date` (that one is problematic because `Date` is also the name of the JS interface)
- `DateTimePicker` => `DateTime`
- `DateField` => `DateField`
- `CalendarPicker` => `CalendarPicker` (or `DatePicker` if we are 100% coherent but that's tricky to give the name of a v5 component)
- `MonthPicker` => `MonthPicker`
- etc...

**Comment by joserodolfofreitas** [comment]
Yes, that would be the most intuitive for me, and we wouldn't need to use the generic "View".

> The DatePicker => Date (that one is problematic because Date is also the name of the JS interface)

I was inspired by retool. But I see your point, thought when it's written as a component doesn't seem that bad.

`<Date pickers: {['month', 'year']} components:{{ monthPicker: MyMonthPicker }} />`

**Comment by flaviendelangle** [comment]
https://codesandbox.io/s/throbbing-forest-n7skkj?file=/src/App.js

The problem is name collision

**Comment by flaviendelangle** [comment]
Follow yesterday's meeting, we decided to wait before applying the changes in order to gather more information about how these Static / View components would be used.

My latest proposal was to use the "Calendar" prefix and to remove any prefix from `Clock`

- MonthPicker => MonthCalendar
- YearPicker => YearCalendar
- DayPicker => DayCalendar
- CalendarPicker => Calendar
- ClockPicker => Clock

This leaves us with the naming of the `Static`, but I feel that once we will have a headless version ready, those wrappers will loose a lot of there interest.
